### PR TITLE
Use class template and semaphore for shared memory

### PR DESCRIPTION
**Pull request by morallo**
Use a class template to define the SharedMem manager class. This allows to re-use exactly the same SharedMemTemplate.h code for sharing data between different hooks. You just need to create specific struct definitions that you pass to the SharedMem Template to create a dedicated manager object.

Instead of storing just a pointer to the heap in shared memory, put the whole shared structure in the memory mapped file for simplicity.

To signal the availability of data from the hook, use a binary semaphore. This is optional, it's up to the producer and consumer to use the IsDataReady() function.

This PR depends on https://github.com/JeremyAnsel/xwa_hooks/pull/9 and https://github.com/Prof-Butts/Hook_XWACockpitLook/pull/20

### CODE REVIEW

File: impl11/ddraw/EffectsRenderer.cpp [hunk 1]
```diff
@@ -981,7 +981,7 @@ void EffectsRenderer::ApplyDiegeticCockpit()
 		// throttle are present.
 		//if (!_bThrottleTransformReady)
 		{
-			float throttle = (DiegeticMesh == DM_THR_ROT_X || DM_THR_ROT_Y || DM_THR_ROT_Z) ?
+			float throttle = (DiegeticMesh == DM_THR_ROT_X || DiegeticMesh == DM_THR_ROT_Y || DM_THR_ROT_Z) ?
```
Comment (Prof-Butts) [inline impl11/ddraw/EffectsRenderer.cpp]: This change looks a bit weird. Why was this changed?

Comment (morallo) [reply impl11/ddraw/EffectsRenderer.cpp]: I tried to fix a (potential) bug I detected thanks to a VS warning but did it only half-way.
In the original code, the conditional expression on the right of of the assignment will always be true because DM_THR_ROT_Y and DM_THR_ROT_Z are not zero.

If you want to do "DiegeticMesh is one of (DM_THR_ROT_X, DM_THR_ROT_Y, DM_THR_ROT_Z), you need to compare with each one explicitly.

If the constants are a bit mask, then you should use the bitwise OR ("|"), but I don't think it's the case.
